### PR TITLE
Adapt RangeIndex for empty merge result with pandas 2.0

### PR DESCRIPTION
**Pull request by phofl**
- [x] xref #9957
- [x] Tests added / passed
- [x] Passes `pre-commit run --all-files`

This would keep the RangeIndex for pandas 2.0 and up instead of the input Index.

The test is somewhat covered from other tests, but I like having an explicit test that is easy to understand.

### CODE REVIEW

File: dask/dataframe/tests/test_multi.py [hunk 1]
```diff
@@ -1152,7 +1152,7 @@ def test_merge_by_index_patterns(how, shuffle_method):
     def fix_index(out, dtype):
         # Workaround pandas bug where output dtype of empty index will be int64
```
Comment (j-bennet) [inline dask/dataframe/tests/test_multi.py:1153]: The comment says "workaround pandas bug". I know it's not your comment, but this is not really a bug, right? it was a choice in pandas? Maybe we can fix the comment too.

Comment (phofl) [reply dask/dataframe/tests/test_multi.py]: Adjusted comments, the expected output was actual object on 1.5.x, so fixed that too

File: dask/dataframe/tests/test_multi.py [hunk 2]
```diff
@@ -1150,9 +1150,9 @@ def test_merge_by_index_patterns(how, shuffle_method):
     pdf7r = pd.DataFrame({"c": [5, 6, 7, 8], "d": [5, 4, 3, 2]}, index=list("fghi"))
 
     def fix_index(out, dtype):
-        # Workaround pandas bug where output dtype of empty index will be int64
-        # even if input was object.
-        if len(out) == 0:
+        # Workaround pandas behvaior where output dtype of empty index will be
```
Comment (j-bennet) [inline dask/dataframe/tests/test_multi.py:1153]:
```suggestion
        # Workaround pandas behavior where output dtype of empty index will be
```

**Review by j-bennet — APPROVED** [review]
👍

**Comment by phofl** [comment]
This is ugly now, main problem:

The Dask data frames meta information don't know that the merge will be empty and so don't know that we end up with a RangeIndex. So Index before and after compute diverges

File: dask/dataframe/multi.py [hunk 3]
```diff
@@ -292,10 +293,10 @@ def merge_chunk(
...
         out = out[result_meta.columns]
 
-    # Workaround pandas bug where if the output result of a merge operation is
-    # an empty dataframe, the output index is `int64` in all cases, regardless
-    # of input dtypes.
-    if len(out) == 0 and empty_index_dtype is not None:
+    # Workaround pandas behavior where if the output result of a merge operation is
+    # an empty dataframe, the output index is object in all cases, regardless
+    # of input dtypes for pandas < 2.0.
+    if len(out) == 0 and empty_index_dtype is not None and not PANDAS_GT_200:
```
Comment (j-bennet) [inline dask/dataframe/multi.py:299]: If we decided we want to match pandas behavior, why not do it every time?

**Review by j-bennet — COMMENTED** [review]
Ooof, this quickly became not pretty. I don't feel so good about matching `pandas` behavior with 2.0 but not below.

If I understand this correctly, index dtype was enforced to be the same as original dtype so that what we have in meta matches what we have in result. Perhaps it's better to remove our previous tweaks, and accept the fact that in some (hopefully rare) cases meta index dtype will disagree with result index dtype. We probably will have to modify `assert_eq` to accept an extra parameter for that.

I'd like to hear what @jrbourbeau thinks about it.

**Comment by phofl** [comment]
I reverted it for now and call compute to see if this passes. Will have to adjust based on what we will decide to do with the meta attribute

**Pull request closed without merging**